zenoh/src/api/bytes.rs

```diff
@@ -819,2538 +712,831 @@ impl std::fmt::Display for ZDeserializeError {
-    type Output = ZBytes;
-
-    fn serialize(self, t: &mut [u8]) -> Self::Output {
-        ZSerde.serialize(&*t)
+    fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+        Ok(self.as_bytes().into())
     }
 }
-
-impl From<&mut [u8]> for ZBytes {
-    fn from(t: &mut [u8]) -> Self {
-        ZSerde.serialize(t)
+impl<'a> Deserialize<'a> for Cow<'a, str> {
+    fn read(_reader: &mut ZBytesReader) -> Result<Self, ZDeserializeError> {
+        unimplemented!()
     }
-}
-
-// Cow<[u8]>
-impl<'a> Serialize<Cow<'a, [u8]>> for ZSerde {
-    type Output = ZBytes;
-
-    fn serialize(self, t: Cow<'a, [u8]>) -> Self::Output {
-        ZBytes::new(t.to_vec())
+    fn deserialize(zbytes: &'a ZBytes) -> Result<Self, ZDeserializeError> {
+        Ok(match <Cow<'a, [u8]>>::deserialize(zbytes)? {
+            Cow::Borrowed(s) => std::str::from_utf8(s).or(Err(ZDeserializeError))?.into(),
+            Cow::Owned(vec) => String::from_utf8(vec).or(Err(ZDeserializeError))?.into(),
+        })
     }
-}
-
-impl From<Cow<'_, [u8]>> for ZBytes {
-    fn from(t: Cow<'_, [u8]>) -> Self {
-        ZSerde.serialize(t)
+    fn try_from_zbytes(zbytes: ZBytes) -> Result<Self, ZDeserializeError> {
+        String::try_from_zbytes(zbytes).map(Cow::Owned)
     }
 }
 
-impl<'a> Serialize<&Cow<'a, [u8]>> for ZSerde {
-    type Output = ZBytes;
-
-    fn serialize(self, t: &Cow<'a, [u8]>) -> Self::Output {
-        ZBytes::new(t.to_vec())
-    }
+// numeric
+macro_rules! impl_num {
+    ($($ty:ty),* $(,)?) => {$(
+        impl SerializedSize for $ty {
+            const FIXED_SIZE: Option<usize> = Some(size_of::<$ty>());
+        }
+        impl<'a> Serialize<'a> for $ty {
+            type Error = Infallible;
+            fn write(self, writer: &mut ZBytesWriter) -> Result<(), Self::Error> {
+                self.to_le_bytes().write(writer)
+            }
+            fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+                unimplemented!()
+            }
+        }
+        impl<'a> Serialize<'a> for &$ty {
+            type Error = Infallible;
+            fn write(self, writer: &mut ZBytesWriter) -> Result<(), Self::Error> {
+                (*self).to_le_bytes().write(writer)
+            }
+            fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+                unimplemented!()
+            }
+        }
+        impl Deserialize<'_> for $ty {
+            fn read(reader: &mut ZBytesReader) -> Result<Self, ZDeserializeError> {
+                <[u8; {size_of::<$ty>()}]>::read(reader).map(<$ty>::from_le_bytes)
+            }
+            fn deserialize(_zbytes: &'_ ZBytes) -> Result<Self, ZDeserializeError> {
+                unimplemented!()
+            }
+            fn try_from_zbytes(_zbytes: ZBytes) -> Result<Self, ZDeserializeError> {
+                unimplemented!()
+            }
+        }
+    )*};
 }
+impl_num!(i8, i16, i32, i64, i128, isize, u8, u16, u32, u64, u128, usize, f32, f64);
```

Comment:
I'm not sure implementation for isize/usize makes sense anymore since their sizes are platform dependent (we could always use vle, but the issue is that in some languages they are not independent types and are defined in terms of other integers).